hbase-protocol-shaded/src/main/protobuf/server/master/Master.proto

```diff
@@ -1284,4 +1307,16 @@ service ClientMetaService {
    * Get current meta replicas' region locations.
    */
   rpc GetMetaRegionLocations(GetMetaRegionLocationsRequest) returns(GetMetaRegionLocationsResponse);
+
+  /**
+   * Get meta region locations for a given row
+   */
+  rpc LocateMetaRegion(LocateMetaRegionRequest)
+    returns(LocateMetaRegionResponse);
+
+  /**
+   * Get all meta regions locations
+   */
+  rpc GetAllMetaRegionLocations(GetAllMetaRegionLocationsRequest)
+    returns(GetAllMetaRegionLocationsResponse);
```

Comment:
ClientMetaService backed ConnectionRegistry?

So, we get one Region or all of them... nothing in-between?

No Scan support here because we want 'dumbed-down API' in ConnectionRegistry or in exposed ClientMetaService? We want to be able to implement hbase:meta Region locations in cache that we can serve from ZK for your ZKConnectionRegistry impl?

Reply:
I think for now it is fine? Later we could add something like prefetch to the protobuf message if we want to load more than 1 regions per call.

Keep the interface simple is important for implementing cache server. In general, if we only support these two methods, we could just use an in memory map to cache the locations(just like waht we have done in client side meta cache), but if we want to support Scan, then probably we can only make use of a whole HRegion implementation...

Reply:
The api does not allow 'paging' through the hbase:master table Regions. Even if we prefetch the API is awkward for getting next Region.

If a paging API, we'd only need one method?

I think current situation where we only get one Region because there is only one makes this API seem ok.

Reply:
There is a locateMeta method, which is to be used at the normal read/write path, to support locating a meta region with specific key. And what I said above, about the prefetch, is to add a prefetch parameter for the locateMeta method, not for the GetAllMetaRegionLocations method.

The GetAllMetaRegionLocations method are mainly designed to have two usages, one is for other cache servers to pull all the content so it can serve the locateMeta requests from client, another is for some maintainance methods, such as RegionLocator.getAllRegionLocations. We should not call this method for in normal read write path.